man/cswc.1
.TH CSWC 1 "11 April 2019"
.\"
.\" (C) Copyright 2003-2019 Diomidis Spinellis
.\"
.\" This file is part of CScout.
.\"
.\" CScout is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" CScout is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with CScout.  If not, see <http://www.gnu.org/licenses/>.
.\"
.SH NAME
cswc \- CScout workspace compiler
.SH SYNOPSIS
\fBcswc\fP 
[\fB\-gv\fP]
[\fB\-d\fP \fIdirectory\fP]
[\fIfile\fR]
.SH DESCRIPTION
\fIcswc\fP is a workspace compiler for the \fICScout\fP C source code analyzer
and refactoring browser.
\fICScout\fP integrates in a single process the functionality of 
a multi-project build engine,
an ANSI C preprocessor, and
the parts of a C compiler up to and including the semantic analysis
based on types.
The build engine functionality is required to allow the user to process 
multiple compilation and link units as a single batch.
Only thus can \fICScout\fP detect dependencies across different files and
projects.
Each compilation unit can reside in a different directory and can require
processing using different macro definitions or a different include file path.
In a normal build process these options are typically specified in a 
\fIMakefile\fP.
The \fICScout\fP operation is similarly guided by a 
declarative workspace definition file.
To decouple the complexity of the \fICScout\fP workspace processing
specification from its actual operation, and to encouriage experimentation
with alternative (e.g. IDE-based) workspace specification mechanisms,
\fICScout\fP is guided by a very simple imperative script typically
generated from more sophisticated workspace definitions by 
\fIcswc\fP, the \fICScout\fP workspace compiler.
.PP
This manual page describes the \fIcswc\fP invocation and command-line
options.
Details about its input and output formats, setup, and configuration can be
found in the online hypertext documentation and at the project's home page
http://www.spinellis.gr/cscout.
.PP
.SH OPTIONS
.IP "\fB\-d\fP \fIdirectory\fP"
Specify the directory to use for locating the \fICScout\fP configuration files.
.IP "\fB\-g\fP"
Compile the project as generic rather than host-specific C code.
This means that the generated files will include CScout's standard-C include
and macro definition files (\fIstdc-incs.h\fP and \fIstdc-defs.h\fP),
rather than the host-specific ones (\fIhost-incs.h\fP and \fIhost-defs.h\fP).
.IP "\fB\-v\fP"
Display  the \fIcswc\fP version and copyright information and exit.
.PP
.SH EXAMPLE
The following is a configuration file used for processing the
\fIapache\fP web server.
.PP
.DS
.ft C
.nf
workspace apache {
    cd "/usr/local/src/apache/src"

    ro_prefix "/usr/local/src/apache/src/include/ap_config"

    # Global project definitions
    define HTTPD_ROOT "/usr/local/apache"
    define SUEXEC_BIN "/usr/local/apache/bin/suexec"
    define SHARED_CORE_DIR "/usr/local/apache/libexec"
    define DEFAULT_PIDLOG "logs/httpd.pid"
    define DEFAULT_SCOREBOARD "logs/httpd.scoreboard"
    define DEFAULT_LOCKFILE "logs/httpd.lock"
    define DEFAULT_XFERLOG "logs/access_log"
    define DEFAULT_ERRORLOG "logs/error_log"
    define TYPES_CONFIG_FILE "conf/mime.types"
    define SERVER_CONFIG_FILE "conf/httpd.conf"
    define ACCESS_CONFIG_FILE "conf/access.conf"
    define RESOURCE_CONFIG_FILE "conf/srm.conf"

    define AUX_CFLAGS
    define LINUX 22 
    define USE_HSREGEX 
    define NO_DL_NEEDED

    # Give project-specific directory and include path properties
    project gen_uri_delims {
        cd "main"
        ipath "../os/unix"
        ipath "../include"
        file gen_uri_delims.c
    }

    # Alternative formulation; specify per-file properties
    project gen_test_char {
        file gen_test_char.c {
            cd "main"
            ipath "../os/unix"
            ipath "../include"
        }
    }

    # httpd executable; specify directory-based properties
    project httpd {
        directory main {
            ipath "../os/unix"
            ipath "../include"
            file alloc.c buff.c http_config.c http_core.c
            file http_log.c http_main.c http_protocol.c
            file http_request.c http_vhost.c util.c util_date.c
            file util_script.c util_uri.c util_md5.c rfc1413.c
        }
        directory regex {
            ipath "."
            ipath "../os/unix"
            ipath "../include"
            define POSIX_MISTAKE
            file regcomp.c regexec.c regerror.c regfree.c
        }
        directory os/unix {
            ipath "../../os/unix"
            ipath "../../include"
            file os.c os-inline.c
        }
        directory ap {
            ipath "../os/unix"
            ipath "../include"
            file ap_cpystrn.c ap_execve.c ap_fnmatch.c ap_getpass.c 
            file ap_md5c.c ap_signal.c ap_slack.c ap_snprintf.c 
            file ap_sha1.c ap_checkpass.c ap_base64.c ap_ebcdic.c
        }
        directory modules/standard {
            ipath "../../os/unix"
            ipath "../../include"
            file mod_env.c mod_log_config.c mod_mime.c
            file mod_negotiation.c mod_status.c mod_include.c
            file mod_autoindex.c mod_dir.c mod_cgi.c mod_asis.c
            file mod_imap.c mod_actions.c mod_userdir.c
            file mod_alias.c mod_access.c mod_auth.c mod_setenvif.c
        }
        directory . {
            ipath "./os/unix"
            ipath "./include"
            file modules.c buildmark.c
        }
    }
}
.ft P
.fi
.DE
.PP
.SH "SEE ALSO"
cscout(1)
.SH AUTHOR
(C) Copyright 2003 Diomidis Spinellis.


src/cswc.pl
#!/usr/bin/env perl
#
# (C) Copyright 2003-2019 Diomidis Spinellis
#
# This file is part of CScout.
#
# CScout is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# CScout is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with CScout.  If not, see <http://www.gnu.org/licenses/>.
#
# Compile a project description into a C-file compilation script
#
#

# Syntax:
#
## Comment
#workspace name {
#	cd "dirname"
#	define foo bar
#	ro_prefix "path1"
#	ro_prefix "path2"
#	ipath "blah"
#	ipath "foobar"
#	readonly
#	readonly "fname"
#	# Linkage unit
#	project name {
#		cd ...
#		define ...
#		ipath ...
#		readonly
#		directory name {
#			readonly
#			file "fname.c" {
#				cd ...
#				define ...
#				ipath ...
#				readonly
#			}
#		}
#		# Or:
#		file fname2.c ...
#	}
#}
#

use Getopt::Std;
if (!getopts('d:gv')) {
	print STDERR "usage: $0 [-gv] [-d directory] [file]\n";
	exit(1);
}

if ($opt_v) {
	print STDERR "cswc - CScout workspace compiler\n\n" .
	# 80 column terminal width----------------------------------------------------
	"(C) Copyright 2003-2019 Diomidis Spinelllis.\n\n" .
	"CScout is distributed as open source software under the GNU General Public\n" .
	"License, available in the CScout documentation and online at\n" .
	"http://www.gnu.org/licenses/.\n" .
	"Other licensing options and professional support are available on request.\n";
	exit(0);
}

# Installation directory:
# Search order
# $opt_d, .cscout, $CSCOUT_HOME, $HOME/.cscout, and
# INSTALL_INCLUDE
use Cwd 'abs_path';

if (defined($opt_d) && -d $opt_d) {
	$instdir = $opt_d;
} elsif (-d '.cscout') {
	$instdir = '.cscout';
} elsif (defined($ENV{CSCOUT_HOME})) {
	$instdir = $ENV{CSCOUT_HOME};
} elsif (defined($ENV{HOME}) && -d $ENV{HOME} . '/.cscout') {
	$instdir = $ENV{HOME} . '/.cscout';
} elsif (-d 'INSTALL_INCLUDE') {
	$instdir = 'INSTALL_INCLUDE';
} else {
	print STDERR "Unable to identify a CScout installation directory\n";
	print STDERR 'Create ./.cscout, or $HOME/.cscout, use -d, or set the $CSCOUT_HOME variable' . "\n";
	exit(1);
}
$instdir = abs_path($instdir);

my ($incs, $defs);

if (defined($opt_g)) {
	$incs = 'stdc-incs.h';
	$defs = 'stdc-defs.h';
} else {
	$incs = 'host-incs.h';
	$defs = 'host-defs.h';
}

if (!-r ($f = "$instdir/$incs")) {
	print STDERR "Unable to read $f: $!\n";
	print STDERR "Create the file in the directory $instdir\nby copying the appropriate compiler-specific file\n";
	exit(1);
}

if (!-r ($f = "$instdir/$defs")) {
	print STDERR "Unable to read $f: $!\n";
	print STDERR "Create the file in the directory $instdir\nby copying the appropriate compiler-specific file\n";
	exit(1);
}

if ($#ARGV == -1) {
	print STDERR "Reading workspace decription from the standard input\n";
}

while (<>) {
	chop;
	s/#.*//;
	s/^\s+//;
	s/\s+$//;
	next if (/^$/);
	if (/(^workspace|project|file|directory)\s+([^ ]+)\s*\{$/) {
		beginunit($1, $2);
	} elsif (/^file\s+(.*[^{])$/) {
		@files = split(/\s+/, $1);
		for $f (@files) {
			beginunit('file', $f);
			endunit();
		}
	} elsif (/^ro_prefix\s+(\".*\")$/) {
		print "#pragma ro_prefix $1\n";
	} elsif (/^cd\s+\"(.*)\"$/) {
		directory($1);
	} elsif (/^define\s+(\w+)$/) {
		$defines{$unit . $name} .= "#define $1\n";
		# print "DEBUG define $1 ($unit $name)\n";
	} elsif (/^define\s+(\w+)\s*(.*)$/) {
		$defines{$unit . $name} .= "#define $1 $2\n";
		# print "DEBUG define $1 $2 ($unit $name)\n";
	} elsif (/^ipath\s+(\".*\")$/) {
		$ipaths{$unit . $name} .= "#pragma includepath $1\n";
	} elsif (/^readonly\s+(\".*\")$/) {
		print "#pragma readonly $1\n";
	} elsif (/^readonly$/) {
		$readonly{$unit . $name} = 1;
	} elsif (/^\}$/) {
		endunit();
	} else {
		print STDERR "Syntax error: $_\n";
	}
}

sub endunit
{
	if ($unit eq 'file') {
		my $i;
		# Include current ones in the scope
		push(@units, $unit);
		push(@names, $name);
		for ($i = 0; $i <= $#units; $i++) {
			my $u = $units[$i];
			my $n = $names[$i];
			print $ipaths{$u . $n};
			print $defines{$u . $n};
			# print "DEBUG looking defines for $u $n\n";
			print "#pragma readonly \"$name\" // $u\n" if ($readonly{$u . $n});
		}
		# Current ones no longer needed
		pop(@units);
		pop(@names);
		print "#include \"$instdir/$incs\"\n";
		print "#pragma process \"$name\"\n\n";
	}
	if (defined($dir{$unit})) {
		print "#pragma echo \"Exiting directory $dir{$unit}\\n\"\n";
		print "#pragma popd\n";
	}
	print "#pragma block_exit\n" unless($unit eq 'workspace' || $unit eq 'directory');
	print "#pragma echo \"Done processing $unit $name\\n\"\n";
	$unit = pop(@units);
	$name = pop(@names);
	$readonly{$unit . $name} = pop(@readonlys);
}


sub beginunit
{
	push(@units, $unit);
	push(@names, $name);
	push(@readonlys, $readonly{$unit . $name});
	($unit, $name) = @_;
	# print "DEBUG beginunit $unit $name\n";
	undef $dir{$unit};
	undef $defines{$unit . $name};
	undef $ipaths{$unit . $name};
	undef $readonly{$unit . $name};
	print "// $unit $name\n";
	print "#pragma echo \"Processing $unit $name\\n\"\n";
	if ($unit eq 'project') {
		print "#pragma project \"$name\"\n";
		print "#pragma block_enter\n";
	} elsif ($unit eq 'directory') {
		directory($name);
	} elsif ($unit eq 'file') {
		print "#pragma block_enter\n";
		print "#pragma clear_defines\n";
		print "#pragma clear_include\n";
		print "#pragma process \"$instdir/$defs\"\n";
	}
}

sub directory
{
	my($dir) = @_;
	$dir{$unit} = $dir;
	print "#pragma echo \"Entering directory $dir\\n\"\n";
	print "#pragma pushd \"$dir\"\n";
}
